File: by-kotlin/kt-boj/solutions/13023-ABCDE.kt
import java.io.BufferedReader
import java.io.InputStreamReader


private lateinit var matrix: Array<Array<Int>>
private lateinit var lists: Array<MutableList<Int>>
private lateinit var edges: MutableList<Array<Int>>

fun solve13023(n: Int): Int {
    for (i in 0 until edges.size) {
        for (j in 0 until edges.size) {
            val (a, b) = edges[i]
            val (c, d) = edges[j]

            if (a == c || a == d || b == c || b == d) continue
            if (matrix[b][c] != 1) continue

            for (e in lists[d]) {
                if (e == a || e == b || e == c || e == d) continue
                return 1
            }
        }
    }
    return 0
}

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    val (n, m) = readLine().split(" ").map { it.toInt() }

    matrix = Array(n) { Array(n) { 0 } }
    lists = Array(n) { mutableListOf() }
    edges = mutableListOf()

    repeat(m) {
        val (a, b) = readLine().split(" ").map { it.toInt() }

        matrix[a][b] = 1
        matrix[b][a] = 1

        lists[a].add(b)
        lists[b].add(a)

        edges.add(arrayOf(a, b))
        edges.add(arrayOf(b, a))
    }
    print(solve13023(n))
}

File: by-python/boj/13023-ABCDE.py
def solve():
    for i in range(len(edges)):
        for j in range(len(edges)):
            A, B = edges[i]
            C, D = edges[j]

            if A == C or A == D or B == C or B == D:
                continue

            if matrix[B][C] != 1:
                continue

            for E in lists[D]:
                if E == A or E == B or E == C or E == D:
                    continue
                print(1)
                return
    print(0)


if __name__ == '__main__':
    N, M = map(int, input().split())
    matrix = [[0] * (N) for _ in range(N)]
    lists = [[] for _ in range(N)]
    edges = []

    for _ in range(M):
        a, b = map(int, input().split())

        matrix[a][b] = 1
        matrix[b][a] = 1

        lists[a].append(b)
        lists[b].append(a)

        edges.append((a, b))
        edges.append((b, a))

    solve()
    # for m in matrix:
    #     print(m)
    # 
    # for l in lists:
    #     print(l)
    # 
    # for e in edges:
    #     print(e)
